[Salesforce] Add Salesforce connector

navarone-feekery opened the pull request:
## Related to https://github.com/elastic/enterprise-search-team/issues/5183

Adds Salesforce as a new connector.
This connector is in beta for 8.10.

- Ingests the SObjects Accounts, Contacts, Leads, Campaigns, Opportunities, Cases
- Ingest ContentDocuments and extracts binary data from their associated files

## Checklists

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [x] Considered corresponding documentation changes
- [x] Contributed any configuration settings changes to the configuration reference

wangch079 commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1475 @@
+
+    async def queryable_sobjects(self):
+        """Cached async property"""
+        if self._queryable_sobjects is not None:
```
This can be simplified by `@cached_property`

navarone-feekery replied:
It's asynchronous though. My understanding is that cached_property doesn't support async calls.
https://asyncstdlib.readthedocs.io/en/stable/source/api/functools.html#async-caches

wangch079 replied:
TIL. Didn't realize that.

navarone-feekery replied:
Sorry I deleted my comment because I think I'm wrong? cached_property doesn't work from functools but the very library I linked has an async cached_property
(linking again) https://asyncstdlib.readthedocs.io/en/stable/source/api/functools.html#async-caches
We don't use this library though so I don't think it's worth introducing. So it IS possible to cache, just not with functools.

wangch079 commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1475 @@
+                params=params,
+            )
+            yield response.get("records")
+            if response.get("done", True) is True:
```
Why defaults to `True`? Will there be scenario where `done` is not returned?

navarone-feekery replied:
I'm just being defensive in my coding. I'm pretty sure there will always be a `done` field.

wangch079 replied:
My point is, should it be defaulted to `False`?

navarone-feekery replied:
If it's defaulted to `False` it will continue trying to get the next page, which might not exist and will raise an error. But maybe that is better, if their API got weird we probably don't want to silently ingest half their dataset.

Alternative: keep it as `True` but log a warning if the `done` field is missing. WDYT?

wangch079 replied:
Or we could ignore it, and just rely on `nextRecordsUrl`.

We stop if there's no `nextRecordsUrl`.

navarone-feekery replied:
That's a lot simpler, I'll do that

wangch079 commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1475 @@
+            url = f"{self.base_url}{response.get('nextRecordsUrl')}"
+            params = None
+
+    async def _execute_non_paginated_query(self, soql_query):
```
I see this is used by `get_case_feeds` only. Why `case_feeds` doesn't need pagination?

navarone-feekery replied:
I was copying from the WS connector and thought (wrongly) that case_feeds were not being paginated (partly because we have a limit on the query of 500). I dug deeper and it _is_ paginated, so I'll fix that.

navarone-feekery replied:
case feeds also support pagination now

wangch079 reviewed (commented):
I see there are classes `SalesforceClient`, `SalesforceAPIToken`, `SalesforceSoqlBuilder`, `SalesforceDocMapper` and `SalesforceDataSource`. I think `SalesforceClient` takes too much responsibility. We should have the duty of each class clearly separated.

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1475 @@
+        skipped_exceptions=[InvalidCredentialsException],
+    )
+    async def get_token(self):
+        while True:
```
Ah missed removing it when I added the retryable, nice catch!

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1475 @@
+        self.salesforce_client.set_logger(self._logger)
+
+    @classmethod
+    def get_default_configuration(cls):
```
required is `True` by default if you don't specify a key

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1475 @@
+        return self._token
+
+    async def generate(self):
+        with self._non_blocking_lock():
```
I may be wrong but I thought our requests were threaded, so there's a chance that multiple requests are attempting to get a token at the same time. Is that not the case?
Looking at it now, the Salesforce connector is not threading anything. I don't know how it works at the client level, though.

navarone-feekery replied:
I looked further into this and I don't think there will be competition, so I've removed the lock.

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1475 @@
+        }
+
+    def token(self):
+        return self._token
```
The token does expire eventually. The user can define the duration for this. Anecdotally my tokens for testing last a few hours but they aren't permanent. I don't know how long a sync of a very large dataset could take, so I'd prefer to be able to "refresh" (i.e. re-get) the token.

wangch079 replied:
TIL. That means the token is fetched without expiry time, and not renewable. So I guess the only option left is trial and error

wangch079 replied:
But still think it might be worth combining methods `token` and `generate`.

navarone-feekery replied:
I've redone this and now there's a single func called `token`.
There's also no need to pre-emptively call it, calling `_auth_headers` will attempt to fetch a token if one doesn't exist yet.

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1475 @@
+    async def _get_json(self, url, params=None):
+        response_body = None
+
+        while True:
```
Same reason as above, will remove 👍🏻

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1475 @@
+            await self.session.close()
+            del self.session
+
+    async def get_docs(self):
```
Moved!

navarone-feekery commented:
@wangch079 Thanks for the review. I implemented most of your feedback. I also moved around a lot of the logic so it wasn't so heavy in `SalesforceClient`.

jedrazb commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1443 @@
+            },
+        }
+
+    async def validate_config(self):
```
Should we add some programatic validations here?  Probably good to check that the config fields are non-null. We should throw an Exception otherwise.

jedrazb commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1443 @@
+    def get_default_configuration(cls):
+        return {
+            "client_id": {
+                "label": "Client ID",
```
Suggestion: add RCF `validations` to configuration fields, [RCF docs](https://github.com/elastic/connectors-python/blob/main/docs/DEVELOPING.md#rich-configurable-fields)

navarone-feekery replied:
What unique validations do you think these require?

navarone-feekery replied:
Of the possible custom validation types (`'less_than' | 'greater_than' | 'list_type' | 'included_in' | 'regex'`) I don't see anything that we could validate on here.

jedrazb replied:
Ahh you are right, I somehow linked `required` property to live in `validations`. 👍

wangch079 commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1443 @@
+                # 400s have detailed error messages in body
+                error_message = response_body.get("error", "No error dscription found.")
+                if error_message == "invalid_client":
+                    raise InvalidCredentialsException(
```
This can be added to `skipped_exceptions`

jedrazb commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1443 @@
+
+    async def get_docs(self, filtering=None):
+        # We collect all content documents and de-duplicate them before downloading and yielding
+        all_content_docs = []
```
> We collect all content documents and de-duplicate them before downloading and yielding

What does collecting all content documents mean exactly here? Is this expected to be a lot for large datasets? Can we cause OOMs with that reference?

navarone-feekery replied:
Every SObject can have a x number of related ContentDocuments, but the same ContentDocument can be attached to multiple SObjects. This collection is so we can compare them before downloading and avoid downloading duplicate files.

> Can we cause OOMs with that reference?
It's possible. This is something we will find out in QA when we can test against a very large dataset. We can exchange the memory usage for higher CPU usage by checking duplicates on the fly, but we deemed it wasn't necessary because 1. we need to see if it's actually a problem, and 2. it's still a beta connector so it wasn't urgent to fix.

[Discussion link](https://github.com/elastic/connectors-python/pull/1451#discussion_r1294483719)

wangch079 commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1443 @@
+        self.order_by = ""
+        self.limit = ""
+
+    def with_id(self):
+        self.fields.append("Id")
```
```suggestion
    def with_id(self):
        self.fields.append("Id")
        return self
```
I was expecting the methods in `SalesforceSoqlBuilder` will return `self`, so that it can be chained:

```python
query_builder = SalesforceSoqlBuilder("CaseFeed")
return query_builder.with_id().with_default_metafields().build()
```

navarone-feekery replied:
I didn't even know this was possible, thanks!

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1443 @@
+        )
+
+    async def ping(self):
+        # TODO ping something of value (this could be config check instead)
```
What config check are you referring to? `validate_config`?
If so, that's called independently by the [sync job runner here](https://github.com/elastic/connectors-python/blob/5c3db0e12e5105018d30df89c5676a59634e19f1/connectors/sync_job_runner.py#L126).

The ping is then handled a [few steps lower](https://github.com/elastic/connectors-python/blob/5c3db0e12e5105018d30df89c5676a59634e19f1/connectors/sync_job_runner.py#L129). Right now the Salesforce ping is just checking if the `domain` that the user registers even exists.

You're right that we can get rid of the comment 👍🏻

jedrazb replied:
For context, I referred to the `config check` mentioned in the comment.

navarone-feekery replied:
I can't even read my own comments, sorry 😅

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1443 @@
+
+    async def get_accounts(self):
+        if not await self._is_queryable("Account"):
+            self._logger.info(
```
Good call!

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1443 @@
+
+        self._sobjects_cache_by_type = {}
+        self._sobjects_cache_by_type["Account"] = await self._prepare_sobject_cache(
+            "Account"
```
The data fetched is from a different endpoint. It made sense in a structural sense to me to include the `Account` cache here instead of from the results of `get_account`. This is where all caching of SObjects is handled. It's easier to mentally track what's going on and is only 1 extra request.

navarone-feekery commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1443 @@
+        query = await self._accounts_query()
+        async for records in self._yield_non_bulk_query_pages(query):
+            for record in records:
+                content_docs = self._parse_content_documents(record)
```
Not sure why I left this actually, it does feel out of place. I've moved it to `SalesforceDataSource`.

navarone-feekery commented:
@jedrazb @wangch079 thanks heaps for the reviews! I've applied some further changes.

jedrazb commented on connectors/sources/salesforce.py:
```diff
@@ -0,0 +1,1453 @@
+            f"ContentDocument.LatestPublishedVersion.{x}"
+            for x in queryable_version_fields
+        ]
+        [f"'{x[1:]}'" for x in TIKA_SUPPORTED_FILETYPES]
```
Is this an unused statement? Not sure if it's a github PR formatting issue but I have an impression that we initialise this array without any reference to it.

navarone-feekery replied:
Ah that is also in the line below it, it's probably a reformatting error. Removed, thanks for spotting that.

jedrazb reviewed (approved):
🚢 ! I expect that the ftests will be added as a follow up PR. Also saw a few `TODO` in comments that hopefully will be fixed before the release
